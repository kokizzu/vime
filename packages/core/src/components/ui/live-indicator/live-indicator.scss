vime-live-indicator {
  /**
   * @prop --live-indicator-color: The color of the live indicator dot.
   */

  display: flex;
  align-items: center;
  font-size: 13px;
  font-weight: bold;
  letter-spacing: 0.6px;
  color: var(--control-color);

  &.hidden {
    display: none;
  }

  & > div.indicator {
    display: inline-block;
    width: 8px;
    height: 8px;
    border-radius: 50%;
    margin-right: 4px;
    background-color: var(--live-indicator-color, red);
  }
}