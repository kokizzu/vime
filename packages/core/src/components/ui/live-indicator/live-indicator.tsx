import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { PlayerProps, PlayerProp } from '../../core/player/PlayerProp';
import { withPlayerContext } from '../../core/player/PlayerContext';

@Component({
  tag: 'vime-live-indicator',
  styleUrl: 'live-indicator.scss',
})
export class LiveIndicator {
  /**
   * @internal
   */
  @Prop() isLive: PlayerProps[PlayerProp.IsLive] = false;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps[PlayerProp.I18N] = {};

  render() {
    return (
      <Host
        class={{
          hidden: !this.isLive,
        }}
      >
        <div class="indicator" />
        {this.i18n.live}
      </Host>
    );
  }
}

withPlayerContext(LiveIndicator, [
  PlayerProp.IsLive,
  PlayerProp.I18N,
]);
