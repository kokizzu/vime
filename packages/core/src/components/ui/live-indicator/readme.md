# vime-live-indicator

This can be used to indicate to the user that the current media is being streamed live.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/live-indicator/live-indicator.png"
  alt="Vime live indicator component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {7} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <!-- ... -->
      <vime-live-indicator></vime-live-indicator>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### Html

```html {7}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <!-- ... -->
      <vime-live-indicator></vime-live-indicator>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### React

```tsx {6,16}
import React from 'react';
import {
  VimePlayer,
  VimeUi,
  VimeControls,
  VimeLiveIndicator,
} from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeControls>
          <VimeLiveIndicator />
        </VimeControls>
      </VimeUi>
    </VimePlayer>
  );
}
```

### Svelte

```html {6,17} title="example.svelte"
<VimePlayer>
  <!-- ... -->
  <VimeUi>
    <!-- ... -->
    <VimeControls>
      <VimeLiveIndicator />
    </VimeControls>
  </VimeUi>
</VimePlayer>

<script lang="ts">
  import {
    VimePlayer,
    VimeUi,
    VimeIcons,
    VimeControls,
    VimeLiveIndicator,
  } from '@vime/svelte';
</script>
```

### Vue

```html {7,18,26} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeControls>
        <VimeLiveIndicator />
      </VimeControls>
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import {
    VimePlayer,
    VimeUi,
    VimeControls,
    VimeLiveIndicator,
  } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeControls,
      VimeLiveIndicator,
    },
  };
</script>
```

## CSS Custom Properties

| Name                     | Description                          |
| ------------------------ | ------------------------------------ |
| `--live-indicator-color` | The color of the live indicator dot. |

## Dependencies

### Used by

- [vime-default-controls](../controls/default-controls)

### Graph

```mermaid
graph TD;
  vime-default-controls --> vime-live-indicator
  style vime-live-indicator fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
