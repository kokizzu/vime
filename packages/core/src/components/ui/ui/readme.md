# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
