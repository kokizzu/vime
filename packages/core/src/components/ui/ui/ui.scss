vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}