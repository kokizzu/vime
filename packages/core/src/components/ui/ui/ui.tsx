import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        {canShowCustomUI && <slot />}
      </Host>
    );
  }
}

openPlayerWormhole(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
