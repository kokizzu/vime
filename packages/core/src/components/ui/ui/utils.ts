import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
