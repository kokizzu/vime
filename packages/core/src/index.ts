import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
