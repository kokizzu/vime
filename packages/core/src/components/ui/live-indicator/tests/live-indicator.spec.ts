import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../ui/tests';
import { LiveIndicator } from '../live-indicator';

let page: SpecPage;
let indicator: HTMLVimeLiveIndicatorElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [LiveIndicator],
    '<vime-live-indicator />',
  ));

  indicator = page.root!.querySelector('vime-live-indicator')!;
});

it('should be structurally sound', () => {
  expect(indicator).toMatchSnapshot();
});

it('should contain text', () => {
  expect(indicator.innerHTML).toContain('LIVE');
});

it('should be hidden by default', () => {
  expect(indicator).toHaveClass('hidden');
});

it('should not be hidden if media is live', async () => {
  expect(indicator).toHaveClass('hidden');
  indicator.isLive = true;
  await page.waitForChanges();
  expect(indicator).not.toHaveClass('hidden');
});
