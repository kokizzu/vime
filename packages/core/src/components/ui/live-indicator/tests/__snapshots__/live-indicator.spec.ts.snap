// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-live-indicator class="hidden">
  <div class="indicator"></div>
  LIVE
</vime-live-indicator>
`;
